rtrlib/pfx/lpfst/lpfst.c

```diff
@@ -8,151 +8,247 @@
-        if(*array == NULL)
-            return -1;
-        (*array)[*len - 1] = root_node->lchild;
-        if(lpfst_get_children(root_node->lchild, array, len) == -1)
-            return -1;
-    }
-
-    if(root_node->rchild != NULL) {
-        *len += 1;
-        *array = realloc(*array, *len * sizeof(struct lpfst_node *));
-        if(*array == NULL)
-            return -1;
-        (*array)[*len - 1] = root_node->rchild;
-        if(lpfst_get_children(root_node->rchild, array, len) == -1)
-            return -1;
-    }
-    return 0;
+	return n->len == mask_len && lrtr_ip_addr_equal(n->prefix, *p);
+}
+
+static void replace_node_data(struct lpfst_node *a, struct lpfst_node *b)
+{
+	a->prefix = b->prefix;
+	a->len = b->len;
+	a->data = b->data;
+}
+
+struct lpfst_node *lpfst_remove(struct lpfst_node *root,
+				const struct lrtr_ip_addr *prefix,
+				const uint8_t mask_len,
+				const unsigned int lvl)
+{
+	/* If the node has no children we can simply remove it */
+	/* If the node has children, we swap the node with the child that
+	 * has the bigger prefix length and drop the child. */
+	if (prefix_is_same(root, prefix, mask_len)) {
+		void *tmp;
+
+		if (lpfst_is_leaf(root)) {
+			deref_node(root);
+			return root;
+		}
+
+		/* swap with the left child and drop the child */
+		if (root->lchild && (!root->rchild ||
+				     root->lchild->len > root->rchild->len)) {
+			tmp = root->data;
+			replace_node_data(root, root->lchild);
+			root->lchild->data = tmp;
+
+			return lpfst_remove(root->lchild,
+					    &root->lchild->prefix,
+					    root->lchild->len, lvl + 1);
+		}
+
+		/* swap with the right child and drop the child */
+		tmp = root->data;
+		replace_node_data(root, root->rchild);
+		root->rchild->data = tmp;
+
+		return lpfst_remove(root->rchild,
+				    &root->rchild->prefix,
+				    root->rchild->len, lvl + 1);
+	}
+
+	if (is_left_child(prefix, lvl)) {
+		if (!root->lchild)
+			return NULL;
+		return lpfst_remove(root->lchild, prefix, mask_len, lvl + 1);
+	}
+
+	if (!root->rchild)
+		return NULL;
+	return lpfst_remove(root->rchild, prefix, mask_len, lvl + 1);
+}
+
+static int append_node_to_array(struct lpfst_node ***ary,
+				unsigned int *len,
+				struct lpfst_node *n)
+{
+	struct lpfst_node **new;
+
+	new = realloc(*ary, *len * sizeof(n));
+	if (!new)
+		return -1;
+
+	*ary = new;
+	(*ary)[*len - 1] = n;
+	return 0;
+}
+
+int lpfst_get_children(const struct lpfst_node *root_node,
+		       struct lpfst_node ***array, unsigned int *len)
+{
+	if (root_node->lchild) {
+		*len += 1;
+		if (append_node_to_array(array, len, root_node->lchild))
+			goto err;
+
+		if (lpfst_get_children(root_node->lchild, array, len) == -1)
+			goto err;
+	}
+
+	if (root_node->rchild) {
+		*len += 1;
+		if (append_node_to_array(array, len, root_node->rchild))
+			goto err;
+
+		if (lpfst_get_children(root_node->rchild, array, len) == -1)
+			goto err;
+	}
+
+	return 0;
+
+err:
+	free(*array);
```

Comment:
> So that lpfst_get_children will recursively allocate memory with each child found, right?!
right :-)

> However, due the recursion free(*array) will be called n-times, with n depending on recursive depth. This is allowed as free(NULL) will do nothing, but it is still an unnecessary function call.
Yes, the only way to not do the free() in that case would be to add an if-condition before. That would be the same than what free() is calling already internally.
The unnecessary function call won't have any noticable impact, maybe even the compiler inlines it. 

> Further, this function fails if realloc fails, which is some kind of memory error (e.g. out-of-mem) - maybe add error/debug message here?
I don't know to do that without too much effort. 
The best way would be to forward meaningful errno values to the caller. That is currently nowhere done :-(.
Also whenever a memory allocation fails, the kernel will log an out of memory message.